File: dashboards/ruby/routes/dlp.rb
# frozen_string_literal: true

# DLP routes — rules, violations, test scan

get '/dlp' do
  require_auth!
  rules_res = api_get('/engine/dlp/rules?orgId=default')
  @rules = rules_res[:status] == 200 ? (rules_res[:body]['rules'] || rules_res[:body]) : []
  @rules = @rules.is_a?(Array) ? @rules : []
  violations_res = api_get('/engine/dlp/violations?orgId=default')
  @violations = violations_res[:status] == 200 ? (violations_res[:body]['violations'] || violations_res[:body]) : []
  @violations = @violations.is_a?(Array) ? @violations : []
  erb :dlp
end

post '/dlp/rules/create' do
  require_auth!
  res = api_post('/engine/dlp/rules', {
    name: params[:name],
    pattern: params[:pattern],
    action: params[:action],
    orgId: 'default'
  })
  if res[:status] < 300
    set_flash('DLP rule created successfully', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to create DLP rule') : 'Failed to create DLP rule', 'danger')
  end
  redirect '/dlp'
end

post '/dlp/rules/:id/delete' do
  require_auth!
  res = api_delete("/engine/dlp/rules/#{params[:id]}")
  if res[:status] < 300
    set_flash('DLP rule deleted successfully', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Failed to delete DLP rule') : 'Failed to delete DLP rule', 'danger')
  end
  redirect '/dlp'
end

post '/dlp/scan' do
  require_auth!
  res = api_post('/engine/dlp/scan', {
    content: params[:content],
    orgId: 'default'
  })
  if res[:status] < 300
    set_flash('Scan complete — check results below', 'success')
  else
    set_flash(res[:body].is_a?(Hash) ? (res[:body]['error'] || 'Scan failed') : 'Scan failed', 'danger')
  end
  redirect '/dlp'
end

File: dashboards/python/routes/dlp.py
"""
DLP routes: manage data loss prevention rules, violations, and scanning.
"""

from flask import Blueprint, render_template, request, session, redirect, url_for, flash
from functools import wraps
from utils.api import api_request

dlp_bp = Blueprint('dlp_bp', __name__)


def login_required(f):
    @wraps(f)
    def decorated(*args, **kwargs):
        if 'token' not in session:
            return redirect(url_for('auth_bp.login'))
        return f(*args, **kwargs)
    return decorated


@dlp_bp.route('/dlp')
@login_required
def dlp():
    rules = api_request('GET', '/engine/dlp/rules?orgId=default')
    violations = api_request('GET', '/engine/dlp/violations?orgId=default')
    return render_template(
        'dlp.html',
        active_page='dlp',
        user=session.get('user', {}),
        rules=rules.get('rules', []),
        violations=violations.get('violations', []),
    )


@dlp_bp.route('/dlp/rules/create', methods=['POST'])
@login_required
def create_rule():
    body = {
        'name': request.form['name'],
        'type': request.form['type'],
        'pattern': request.form['pattern'],
        'action': request.form.get('action', 'block'),
        'severity': request.form.get('severity', 'high'),
    }
    result = api_request('POST', '/engine/dlp/rules', body)
    flash('DLP rule created!' if 'id' in result else result.get('error', 'Failed'))
    return redirect(url_for('dlp_bp.dlp'))


@dlp_bp.route('/dlp/rules/<id>/delete', methods=['POST'])
@login_required
def delete_rule(id):
    api_request('DELETE', f'/engine/dlp/rules/{id}')
    flash('DLP rule deleted')
    return redirect(url_for('dlp_bp.dlp'))


@dlp_bp.route('/dlp/scan', methods=['POST'])
@login_required
def scan():
    body = {'content': request.form['content']}
    result = api_request('POST', '/engine/dlp/scan', body)
    flash(f'Scan complete — {len(result.get("violations", []))} violation(s) found')
    return redirect(url_for('dlp_bp.dlp'))
